src/__tests__/parsers/babel_parser_plugin.test.js

```diff
@@ -0,0 +1,50 @@
+import * as babelParser from '@babel/parser';
+import {plugins} from '../../parsers/babel_parser';
+import parse from '../../parsers/index';
+
+const additionalPlugins = [
+  'classPrivateProperties',
+  'classPrivateMethods',
+  'doExpressions',
+  'dynamicImport',
+  'exportDefaultFrom',
+  'exportNamespaceFrom',
+  'functionBind',
+  'functionSent',
+  'importMeta',
+  'logicalAssignment',
+  'nullishCoalescingOperator',
+  'numericSeparator',
+  'objectRestSpread',
+  'optionalCatchBinding',
+  'optionalChaining',
+  'partialApplication',
+  'throwExpressions',
+  'topLevelAwait',
+];
+
+const codeTobeParsed = 'const bla = 1';
+
+const parserConfigFlow = {
+  sourceType: 'module',
+  plugins: [...plugins, ...additionalPlugins, 'flow'],
+};
+
+const parserConfigTypescript = {
+  sourceType: 'module',
+  plugins: [...plugins, ...additionalPlugins, 'typescript'],
+};
+
+describe('test plugins', () => {
+  it('created plugins for javascript file correctly', () => {
+    const spy = jest.spyOn(babelParser, 'parse');
+    parse('fileName.js', codeTobeParsed, false, additionalPlugins);
+    expect(spy).toBeCalledWith(codeTobeParsed, parserConfigFlow);
+  });
+
+  it('created plugins for typescript file correctly', () => {
+    const spy = jest.spyOn(babelParser, 'parse');
+    parse('fileName.ts', codeTobeParsed, false, additionalPlugins);
+    expect(spy).toBeCalledWith(codeTobeParsed, parserConfigTypescript);
+  });
+});
```

Comment:
no you didn't missed them, i have not added any.
i thought about this topic, normally you would only test your package, not the 3rd party depdenencies the project is using.
the plugin features should already have been tested in the babel project.

i mean we could not even fix a issue, if some tests were red, since its not within this project.

Reply:
I see where you are coming from. I think I am less concerned about "if babel implemented the plugin correctly" but "if the plugin list we passed in has the effect we anticipated"...

I played with the new `babel-parser_plugin.test.js` to try to answer my own question:
- comment out all additional plugins, it passed; ok, maybe all of our code fixures/snippets are just standard js that requires no additional plugin...?
- passed in a garbage plugin, it still passed. ok, maybe babel load plugin like a best-effort...?
- I then created some optional chaining code-snippet with empty additional plugin list, thinking for sure it should fail without passing the correct plugin, right? But surprise surprise, it still passed.

At this point, I feel there is something I don't understand or misunderstand about this plugins thing... maybe it is reading the babel.config.js locally? does that mean it will be impacted by users' env as well once we deployed, which might be good, or bad... How do tools like `vscode-jest` or yours decide what plugin list to pass in? If they can't decide then should all users now need to pass in their own list, even if they don't use babel? With the current implementation (minimal default), how disruptive the next release will be? Do we really need to get tangled with the rather complex user babel config/env?

Anyway, tests mean to surface issues, and I think without the simple additional tests I would not have discovered the unexpected behaviors and questions... in addition to tests, we should answer the questions above to make sure we fully understand what the community and users might face with these changes... thoughts?

Reply:
i don't have an answer to your question yet but i appreciate that you try to understand it properly.  so please don't get my comment wrong.

just one side thought, which might be interesting (in terms of testing)

majestic doesn't seem to take this topic this seriously:
https://github.com/Raathigesh/majestic/blob/master/server/services/ast/parser.ts#L10

they just use 3 plugins => "jsx", "classProperties", "optionalChaining"

Reply:
Hey @connectdotz.  I had a look at this too and think that I've made some headway.  Have a look at this [babel documentation](https://babeljs.io/docs/en/next/babel-parser.html#plugins) that I found helpful (I thought I knew what the plugins did by the names - I wasn't always right).  That page will give you some useful snippets to test with later on.

Experimenting with `babel-parser_plugin.test.js`:
Hopefully I'm not pointing out the obvious, but this file is aimed at testing whether the correct plugin config is passed to the parser rather than whether that configured parser can parse the code correctly.  You can see this by the fact that the "code" it is asked to parse is on line 26:
```js
const codeTobeParsed = 'const bla = 1';
```
which as you can see is pretty simple JS and is not testing any of the experimental language features that the plugins are supposed to be supporting.  Which is probably why when you removed all the plugins, it still parsed correctly and the tests passed.

When I extended that line with snippets from the Babel documentation like this:
```js
const codeTobeParsed = `
//classPrivateProperties
class A { #b = 1; }

//classPrivateMethods
class B { #c() {} }

//doExpressions
var a = do { if (true) { 'hi'; } };
`;
```
I got test failures when the appropriate plugins were not enabled.  Hope that helps.

Reply:
exacly @rossknudsen this test was only to verfiy if the addtionalPlugins have been merged with the default plugins correctly.

since i have added this feature i thought it make sense to test it.

Reply:
also did some research on this topic. 

```babel-parser_plugin.test.js``` is a good place to test the babel plugins.

i found that for example ObjectRestSpread works even if no plugin is activated. While classPrivateProperties or doExpressions do only work with plugins actived.

that is probably because some proposal are already on a later stage and already work even without explicitly activating them.

ObjectRestSpread and OptionalChaining are already in stage 4.

does this answer your concerns @connectdotz ?

Reply:
I dig into the babel-parser code today and found that it checked for a small set (~11) of plugins only, such as `logicalAssignment`, `classProperties`, and the document did mention

> The latest ECMAScript version enabled by default (ES2017).

That explained why some features, such as optionalChaining, objectRestSpread worked without passing in the specific plugins. This is great and does eliminate my biggest concern. Obviously, as the language evolves the extra plugin list will too, but as long as we keep the `babel-parser` up to date we should be able to support the most common use case without passing in any additional plugin...

Let's capture this info in the README so our users don't need to go through the same soul searching process...